### PR TITLE
Added functional vector permutation

### PR DESCRIPTION
**Pull request by guilhermehas**
Added permutation of functional vectors using finite sets permutation.

### CODE REVIEW

**Comment by MatthewDaggitt** [comment]
Thanks for the PR @guilhermehas!

The general consensus between the maintainers at the moment, is that when adding new definitions we prefer to have at least a few lemmas showing that the definition is correct and relatively easy to work with. Would it be possible to add a few?

**Comment by guilhermehas** [comment]
I have added a few theorems.

**Comment by jamesmckinna** [comment]
So, under the`*.Relation.Binary.*` hierarchy for each `Data`/functorial container type, the library distinguishes between `Properties.Propositional`, where the underlying relation on the carrier type is`PropositionalEquality`, and `Properties.Setoid` for an arbitrary `Setoid`/`isEquivalence`relation on the carrier... with the `Propositional` version derived from the `Setoid` by specialising to `PropositionalEquality`...

If I understand this development properly, what you have is the `Propositional` version. How hard would it be to extend your development to an arbitrary `Setoid`? 

Secondly, it might also be nice to know more than just that `Propositional` lists-modulo-permutation is, indeed, an equivalence relation... such as some *concrete* properties regarding eg. `reverse`, `snoc`, flipped `append`... etc

**Comment by guilhermehas** [comment]
I didn't develop for setoids and relations, not because it is hard, but because it does not make sense.
When dealing with permutations using Vectors, we care a lot more about their indices than the content of the vectors. In addition, it is better to prove `xs ++ (ys ++ zs) = (xs ++ ys) ++ zs` than  `xs ++ (ys ++ zs) ~ (xs ++ ys) ++ zs` because the second theorem can be derived by the first one. So in most cases, propositional equality will be better than setoids when using permutations using finite sets.

**Comment by gallais** [comment]
I think james was suggesting an alternative definition that is not based on permutations on fins.
e.g. http://agda.github.io/agda-stdlib/Data.List.Relation.Binary.Permutation.Homogeneous.html#905

**Comment by guilhermehas** [comment]
I think that this definition can be more intuitive than this propositional definition.
I can prove later that both definitions are the same.

File: src/Data/Vec/Functional/Relation/Binary/Permutation/Properties.agda [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+open ≡-Reasoning
+
+private
+  variable
+    ℓ : Level
+    A : Set ℓ
+    n : ℕ
+    xs ys zs : Vector A n
+
+↭-refl : xs ↭ xs
```
Comment (Taneb) [inline src/Data/Vec/Functional/Relation/Binary/Permutation/Properties.agda:29]: Could you use the definitions in `Relation.Binary.Indexed.Heterogeneous.Definitions` in this type and the types of the other definitions in this module?

File: src/Data/Vec/Functional/Relation/Binary/Permutation/Properties.agda [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+-- Properties of permutation
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Data.Vec.Functional.Relation.Binary.Permutation.Properties where
+
+open import Level using (Level)
+open import Data.Product using (_,_; proj₁; proj₂)
+open import Data.Nat using (ℕ)
```
Comment (Taneb) [inline src/Data/Vec/Functional/Relation/Binary/Permutation/Properties.agda:13]: Tighten imports

```suggestion
open import Data.Product.Base using (_,_; proj₁; proj₂)
open import Data.Nat.Base using (ℕ)
```

File: src/Data/Vec/Functional/Relation/Binary/Permutation.agda [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+--
+-- Permutation relations over Vector
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Data.Vec.Functional.Relation.Binary.Permutation where
+
+open import Level using (Level)
+open import Data.Product using (Σ-syntax)
```
Comment (Taneb) [inline src/Data/Vec/Functional/Relation/Binary/Permutation.agda:12]: Tighten imports
```suggestion
open import Data.Product.Base using (Σ-syntax)
```

File: src/Data/Vec/Functional/Relation/Binary/Permutation.agda [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+open import Data.Fin.Permutation using (Permutation; _⟨$⟩ʳ_)
+open import Data.Vec.Functional using (Vector)
+open import Relation.Binary.Indexed.Heterogeneous.Core using (IRel)
+open import Relation.Binary.PropositionalEquality using (_≡_)
```
Comment (Taneb) [inline src/Data/Vec/Functional/Relation/Binary/Permutation.agda:16]: Tighten imports
```suggestion
open import Relation.Binary.PropositionalEquality.Core using (_≡_)
```

**Review by Taneb — CHANGES_REQUESTED** [review]
I've made some minor suggestions. I think this is a good addition to the library.

File: src/Data/Vec/Functional/Relation/Binary/Permutation/Properties.agda (hunk 1 above)
Comment (guilhermehas) [reply src/Data/Vec/Functional/Relation/Binary/Permutation/Properties.agda]: I added these definitions.

Comment (Taneb) [reply src/Data/Vec/Functional/Relation/Binary/Permutation/Properties.agda]: Could you change the changelog to use them too?

**Review by Taneb — APPROVED** [review]

**Review by JacquesCarette — APPROVED** [review]

**Review by gallais — APPROVED** [review]

**Pull request merged**